File: src/main/java/de/melsicon/talk/value/simple/immutables/EmailAddress.java
package de.melsicon.talk.value.simple.immutables;

import com.google.common.base.Preconditions;
import com.google.errorprone.annotations.Immutable;
import org.apache.commons.validator.routines.EmailValidator;
import org.immutables.value.Value;

/** A (validated) e-mail address. */
@Immutable
@Value.Style(passAnnotations = {Immutable.class})
@Value.Immutable
public abstract class EmailAddress {
  private static final EmailValidator EMAIL_VALIDATOR = EmailValidator.getInstance();

  /* package */ EmailAddress() {}

  /**
   * Creates an e-mail address.
   *
   * @param address An address
   * @return An e-mail address
   */
  public static EmailAddress of(String address) {
    return builder().address(address).build();
  }

  /**
   * Creates an e-mail address builder.
   *
   * @return A new builder
   */
  public static ImmutableEmailAddress.Builder builder() {
    return ImmutableEmailAddress.builder();
  }

  /**
   * The e-mail address.
   *
   * @return The address.
   */
  public abstract String address();

  @Value.Check
  /* package */ void check() {
    var address = address();
    Preconditions.checkState(
        EMAIL_VALIDATOR.isValid(address), "%s is not a valid e-mail address", address);
  }
}

File: src/main/kotlin/de/melsicon/talk/value/simple/kotlin/EmailAddress.kt
package de.melsicon.talk.value.simple.kotlin

import org.apache.commons.validator.routines.EmailValidator

val EMAIL_VALIDATOR: EmailValidator = EmailValidator.getInstance()

/**
 * A (validated) e-mail address.
 *
 * @param[address] Set the e-mail address.
 */
data class EmailAddress(val address: String) {
    init {
        require(EMAIL_VALIDATOR.isValid(address)) { "$address is not a valid e-mail address" }
    }
}
